### ISSUE TITLE
Question about `REWARD_WEIGHT_MODE="teacher_p"` and forward KL in OPD

### ISSUE THREAD

**Issue by Ppp-Ttt** [Thinking-Space/Rethinking-OPD#21]
Hi, I have a question about the KL direction when using the following configuration:

```python
TOP_K = 16
TOP_K_STRATEGY = "only_stu"
REWARD_WEIGHT_MODE = "student_p" / "teacher_p"
```

My understanding is:

When `TOP_K_STRATEGY="only_stu"` and `REWARD_WEIGHT_MODE="student_p"`, the objective corresponds to the Reverse KL form used in standard OPD:

```text
p_s * log(p_s / p_t)
```

However, when `REWARD_WEIGHT_MODE="teacher_p"`, it seems to correspond to:

```text
p_t * log(p_s / p_t)
```

rather than the Forward KL version of OPD:

```text
p_t * log(p_t / p_s)
```

Is this understanding correct?

If I want to modify the implementation to use the forward KL form, would the simplest change be in:

```text
verl/verl/workers/actor/dp_actor.py
```

around lines 522– where the current logic is:

```python
        def compute_reward_weights(S_logp, T_logp, valid_mask, weight_mode, normalize=True):
            """Compute weights for reward calculation.
            
            Args:
                S_logp: Student log probabilities (batch, seq, K)
                T_logp: Teacher log probabilities (batch, seq, K)
                valid_mask: Boolean mask for valid tokens (batch, seq, K)
                weight_mode: "student_p", "teacher_p", or "none"
                normalize: If True, apply softmax normalization across K dim.
                          If False, use raw probabilities (masked by valid_mask).
            
            Returns:
                Weights (batch, seq, K)
            """
            if weight_mode == "student_p":
                log_probs = S_logp
            elif weight_mode == "teacher_p":
                log_probs = T_logp
            elif weight_mode == "none":
                # 对于"none"模式，使用均匀分布
                log_probs = torch.zeros_like(S_logp)
            else:
                raise ValueError(f"Unknown reward_weight_mode: {weight_mode}")
            
            log_probs = torch.where(valid_mask, log_probs, torch.full_like(log_probs, -float('inf')))
            
            if normalize:
                norm_log_weights = log_probs - torch.logsumexp(log_probs, dim=-1, keepdim=True)
                weights = torch.exp(norm_log_weights)
            else:
                weights = torch.exp(log_probs)
            
            weights = torch.nan_to_num(weights, nan=0.0, posinf=0.0, neginf=0.0)
            
            return weights
```

change to:

```python
        def compute_reward_weights(S_logp, T_logp, valid_mask, weight_mode, normalize=True):
            """Compute weights for reward calculation.
            
            Args:
                S_logp: Student log probabilities (batch, seq, K)
                T_logp: Teacher log probabilities (batch, seq, K)
                valid_mask: Boolean mask for valid tokens (batch, seq, K)
                weight_mode: "student_p", "teacher_p", or "none"
                normalize: If True, apply softmax normalization across K dim.
                          If False, use raw probabilities (masked by valid_mask).
            
            Returns:
                Weights (batch, seq, K)
            """
            if weight_mode == "student_p":
                log_probs = S_logp
            elif weight_mode == "teacher_p":
                log_probs = T_logp
            elif weight_mode == "none":
                # 对于"none"模式，使用均匀分布
                log_probs = torch.zeros_like(S_logp)
            else:
                raise ValueError(f"Unknown reward_weight_mode: {weight_mode}")
            
            log_probs = torch.where(valid_mask, log_probs, torch.full_like(log_probs, -float('inf')))
            
            if normalize:
                norm_log_weights = log_probs - torch.logsumexp(log_probs, dim=-1, keepdim=True)
                weights = torch.exp(norm_log_weights)
            else:
                weights = torch.exp(log_probs)
            
            weights = torch.nan_to_num(weights, nan=0.0, posinf=0.0, neginf=0.0)

            # new
            if weight_mode == "teacher_p":
                return -weights

            return weights
```

for the Forward KL OPD case?

Could you please confirm whether this is the intended behavior, or whether I am misunderstanding how `REWARD_WEIGHT_MODE="teacher_p"` is implemented?

Thanks!

**Comment by Turingzero0** [comment]
I think your understanding is correct. In our initial implementation, we used `-weights` for reweighting, aiming to make the final loss resemble a forward-KL objective. However, for reasons that are still unclear, this variant performed very poorly in practice.

<img width="493" height="305" alt="Image" src="https://github.com/user-attachments/assets/777f876e-a490-46b4-994e-eb602c2fe1b7" />

<img width="491" height="311" alt="Image" src="https://github.com/user-attachments/assets/a9d89d2b-2cc0-4fba-959d-cb7ce8967adc" />

<img width="487" height="301" alt="Image" src="https://github.com/user-attachments/assets/ad681315-eda3-4c3b-aeae-2368d1bf39a8" />

The purple curve uses `student_p` weighting, the gray curve uses `teacher_p` weighting, and the green curve uses `-teacher_p` weighting. The intuition behind this choice is that weighting by $$-p_{tch}$$ is equivalent to adding a negative sign to the log term, yielding $$\sum p_{tch}\log \frac{p_{tch}}{p_{stu}}$$, which corresponds to minimizing the forward KL. As shown, `teacher_p` weighting does have some effect, but the improvement is generally limited. Using the negative weighting performs poorly.

<img width="491" height="310" alt="Image" src="https://github.com/user-attachments/assets/4ee621fc-24bb-4d22-9aae-2ea970eb9502" />

<img width="495" height="312" alt="Image" src="https://github.com/user-attachments/assets/153f8aeb-ae4c-4f43-8028-23245f4d396e" />

As shown, after adding the negative sign, the loss does indeed become positive (it is consistent with forward KL >0 ), but it is unclear why the training becomes unstable and diverges. I suspect this may be attributed to the issue of implementing KL through the reward (that is KL in reward) , but the exact reason probably requires deriving the formula in detail.

**Comment by Ppp-Ttt** [comment]
I also observed the gradient explosion phenomenon caused by ```-weights```
<img width="918" height="538" alt="Image" src="https://github.com/user-attachments/assets/8602c80e-30be-4798-a9f5-19f7c27b6cf9" />

In fact, I think it is caused by the the gradient generated by this implementation is different from the gradient of the FKL constraint:

**Standard Reverse KL Loss:**\
$$\nabla_\theta \text{KL}\left(p_s\Vert p_t \right)== \sum_k p_s\log\tfrac{p_s}{p_t}\nabla\log p_s $$
we need $$\text{rmscore}=-p_s\log\tfrac{p_s}{p_t}$$ in ppo-style OPD

**Right Reverse KL implement by ```student_p```:**
```
if weight_mode == "student_p":
    log_probs = S_logp
weights = torch.exp(log_probs)
kl_val = S_logp - T_on_S
rm_scores = -kl_val * norm_weights
```
in code, it has been correctly implemented

---

**Standard Forward KL Loss:**\
$$\nabla_\theta \text{KL}\left(p_t\Vert p_s \right)== - \sum_k p_t\nabla\log p_s$$
we need  $$\text{rmscore}=p_t$$

**Wrong Forward KL implement by original ```teacher_p```:**
```
if weight_mode == "teacher_p":
    log_probs = T_logp
weights = torch.exp(log_probs)
kl_val = S_logp - T_on_S
rm_scores = -kl_val * weights
```
where $$\text{rmscore}=-p_t\log\tfrac{p_s}{p_t}$$

**Wrong Forward KL implement by ```-weights```:**
```
if weight_mode == "teacher_p":
    log_probs = T_logp
weights = - torch.exp(log_probs)
kl_val = S_logp - T_on_S
rm_scores = -kl_val * weights
```
where $$\text{rmscore}=-p_t\log\tfrac{p_t}{p_s}$$

**Right Forward KL implement:**
```
if weight_mode == "teacher_p":
    log_probs = T_logp
weights =  torch.exp(log_probs)
rm_scores =  weights
```
where $$\text{rmscore}=p_t$$

If we need forward KL for OPD in ppo-style, we need to set ```rm_scores =   torch.exp(T_logp)```

**Comment by Turingzero0** [comment]
I think your derivation is completely correct. When taking the derivative of the forward KL, $$p_t$$ is treated as a constant, so the KL-in-reward implementation should set `rm_score = p_t`.

Thank you very much for the derivation.
